src/CommitmentsV2/SFA.DAS.CommitmentsV2/Application/Commands/StopApprenticeship/StopApprenticeshipCommandHandler.cs

```diff
@@ -0,0 +1,110 @@
+﻿using MediatR;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using NServiceBus;
+using SFA.DAS.CommitmentsV2.Authentication;
+using SFA.DAS.CommitmentsV2.Data;
+using SFA.DAS.CommitmentsV2.Domain.Exceptions;
+using SFA.DAS.CommitmentsV2.Messages.Commands;
+using SFA.DAS.CommitmentsV2.Messages.Events;
+using SFA.DAS.CommitmentsV2.Shared.Interfaces;
+using SFA.DAS.CommitmentsV2.Types;
+using SFA.DAS.Encoding;
+using SFA.DAS.NServiceBus.Services;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.CommitmentsV2.Application.Commands.StopApprenticeship
+{
+    public class StopApprenticeshipCommandHandler : AsyncRequestHandler<StopApprenticeshipCommand>
+    {
+        private readonly Lazy<ProviderCommitmentsDbContext> _dbContext;
+        private readonly ICurrentDateTime _currentDate;
+        private readonly IAuthenticationService _authenticationService;
+        private readonly IEventPublisher _eventPublisher;
+        private readonly IMessageSession _nserviceBusContext;
+        private readonly IEncodingService _encodingService;
+        private readonly ILogger<StopApprenticeshipCommandHandler> _logger;
+        private const string StopNotificationEmailTemplate = "ProviderApprenticeshipStopNotification";
+
+        public StopApprenticeshipCommandHandler(
+            Lazy<ProviderCommitmentsDbContext> dbContext,
+            ICurrentDateTime currentDate,
+            IEventPublisher eventPublisher,
+            IAuthenticationService authenticationService,
+            IMessageSession nserviceBusContext,
+            IEncodingService encodingService,
+            ILogger<StopApprenticeshipCommandHandler> logger)
+        {
+            _dbContext = dbContext;
+            _currentDate = currentDate;
+            _eventPublisher = eventPublisher;
+            _authenticationService = authenticationService;
+            _nserviceBusContext = nserviceBusContext;
+            _encodingService = encodingService;
+            _logger = logger;
+        }
+
+        protected async override Task Handle(StopApprenticeshipCommand command, CancellationToken cancellationToken)
+        {
+            try
+            {
+                _logger.LogInformation($"Begin stopping apprenticeShip. Apprenticeship-Id:{command.ApprenticeshipId}");
+
+                var party = _authenticationService.GetUserParty();
+                CheckPartyIsValid(party);
+
+                var apprenticeship = await _dbContext.Value.Apprenticeships
```

Comment:
There's a handy extension method on the context called `GetApprenticeshipAggregate` that will get you the `apprenticeship` object with all the entities that make up the aggregate.

Reply:
I've tweaked to use this extension method. Saw some funky results in AT where some of the ones I selected had no account legal entity, so didn't return an apprenticeship. In addition, the query seemed to specify isApproved = 1 all the time but couldn't see where that came from.